File: sdks/ts/memory/src/ingest/state-machine.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Pipeline state machine implemented as plain TypeScript. Defines the
 * formal stage transitions for the ingestion pipeline with retry logic
 * and dead letter handling. No external dependencies.
 *
 * Stages: received -> stored -> chunked -> embedded -> indexed
 * Any stage may transition to "dead_letter" when retries are exhausted.
 */

/** Pipeline stages representing document processing progress. */
export type PipelineStage =
  | 'received'
  | 'stored'
  | 'chunked'
  | 'embedded'
  | 'indexed'
  | 'dead_letter'

/** Ordered stages for validation (excludes dead_letter). */
export const STAGE_ORDER: readonly PipelineStage[] = [
  'received',
  'stored',
  'chunked',
  'embedded',
  'indexed',
]

/** Ordinal position of each stage for comparison. */
const STAGE_ORDINAL: ReadonlyMap<PipelineStage, number> = new Map([
  ['received', 0],
  ['stored', 1],
  ['chunked', 2],
  ['embedded', 3],
  ['indexed', 4],
])

/** Valid single-step forward transitions. */
const VALID_TRANSITIONS: ReadonlyMap<PipelineStage, PipelineStage> = new Map([
  ['received', 'stored'],
  ['stored', 'chunked'],
  ['chunked', 'embedded'],
  ['embedded', 'indexed'],
])

/**
 * Reports whether moving from current to target is a valid single-step
 * forward transition.
 */
export const isValidTransition = (current: PipelineStage, target: PipelineStage): boolean => {
  return VALID_TRANSITIONS.get(current) === target
}

/** Persisted state for a single document in the pipeline. */
export type PipelineStateEntry = {
  readonly documentHash: string
  readonly stage: PipelineStage
  readonly retryCount: number
  readonly lastError: string
  readonly createdAt: string
  readonly updatedAt: string
}

/** Persistence interface for pipeline state entries. */
export type PipelineStateStore = {
  readonly load: (documentHash: string) => Promise<PipelineStateEntry | undefined>
  readonly save: (entry: PipelineStateEntry) => Promise<void>
  readonly listIncomplete: () => Promise<readonly PipelineStateEntry[]>
}

/** Callback invoked on every state transition for observability. */
export type TransitionCallback = (
  documentHash: string,
  from: PipelineStage,
  to: PipelineStage,
  event: string,
) => void

/** Maximum retry count before moving to the dead_letter stage. */
const MAX_DEFAULT_RETRIES = 3

/** Configuration for creating a PipelineStateMachine instance. */
export type PipelineStateMachineConfig = {
  readonly stateStore: PipelineStateStore
  readonly maxRetries?: number
  readonly onTransition?: TransitionCallback
}

/** V1 pipeline state entry from P0-1 for migration support. */
export type V1PipelineStateEntry = {
  readonly documentId: string
  readonly hash: string
  readonly stage: 'stored' | 'chunked' | 'embedded' | 'indexed'
  readonly updatedAt: string
  readonly chunkCount?: number
}

/** Map from V1 stage names to V2 PipelineStage. */
const V1_STAGE_MAP: ReadonlyMap<string, PipelineStage> = new Map([
  ['stored', 'stored'],
  ['chunked', 'chunked'],
  ['embedded', 'embedded'],
  ['indexed', 'indexed'],
])

/**
 * Migrates a V1 pipeline state entry to the V2 format used by the
 * state machine. V1 entries from P0-1 lack retry tracking and use a
 * different type shape.
 */
export const migrateFromV1 = (v1: V1PipelineStateEntry): PipelineStateEntry => {
  const stage = V1_STAGE_MAP.get(v1.stage) ?? 'received'
  return {
    documentHash: v1.hash,
    stage,
    retryCount: 0,
    lastError: '',
    createdAt: v1.updatedAt,
    updatedAt: v1.updatedAt,
  }
}

/**
 * Creates a pipeline state machine manager that orchestrates stage
 * transitions with persistence and observability. Uses a plain
 * switch-based approach matching the Go implementation pattern.
 */
export const createPipelineStateMachine = (config: PipelineStateMachineConfig) => {
  const maxRetries = config.maxRetries ?? MAX_DEFAULT_RETRIES

  /**
   * Resolves the next stage for a forward advance from the current stage.
   * Returns the current stage if no valid forward transition exists.
   */
  const nextStage = (current: PipelineStage): PipelineStage => {
    return VALID_TRANSITIONS.get(current) ?? current
  }

  /**
   * Determines whether the current stage is at or past the target
   * in the forward chain.
   */
  const isAtOrPast = (current: PipelineStage, target: PipelineStage): boolean => {
    const currentOrd = STAGE_ORDINAL.get(current)
    const targetOrd = STAGE_ORDINAL.get(target)
    if (currentOrd === undefined || targetOrd === undefined) return false
    return currentOrd >= targetOrd
  }

  /**
   * Advance a document to the next stage. Validates the transition,
   * persists the new state, and invokes the transition callback.
   */
  const advance = async (
    documentHash: string,
    targetStage: PipelineStage,
  ): Promise<PipelineStateEntry> => {
    const existing = await config.stateStore.load(documentHash)
    const now = new Date().toISOString()

    const currentStage: PipelineStage = existing?.stage ?? 'received'

    // Terminal state: no transitions allowed.
    if (currentStage === 'dead_letter') {
      return existing ?? {
        documentHash,
        stage: 'dead_letter',
        retryCount: 0,
        lastError: '',
        createdAt: now,
        updatedAt: now,
      }
    }

    // Idempotent: already at or past the target.
    if (isAtOrPast(currentStage, targetStage)) {
      return existing ?? {
        documentHash,
        stage: currentStage,
        retryCount: 0,
        lastError: '',
        createdAt: now,
        updatedAt: now,
      }
    }

    // Validate: only single-step forward transitions.
    const next = nextStage(currentStage)
    if (next !== targetStage) {
      return existing ?? {
        documentHash,
        stage: currentStage,
        retryCount: 0,
        lastError: '',
        createdAt: now,
        updatedAt: now,
      }
    }

    const entry: PipelineStateEntry = {
      documentHash,
      stage: next,
      retryCount: existing?.retryCount ?? 0,
      lastError: existing?.lastError ?? '',
      createdAt: existing?.createdAt ?? now,
      updatedAt: now,
    }

    await config.stateStore.save(entry)

    if (config.onTransition !== undefined && currentStage !== next) {
      config.onTransition(documentHash, currentStage, next, 'advance')
    }

    return entry
  }

  /**
   * Record a failure for the document. Increments the retry counter.
   * If retries are exhausted, moves to dead_letter.
   */
  const recordFailure = async (
    documentHash: string,
    error: string,
  ): Promise<PipelineStateEntry> => {
    const existing = await config.stateStore.load(documentHash)
    const now = new Date().toISOString()
    const currentStage: PipelineStage = existing?.stage ?? 'received'
    const currentRetry = existing?.retryCount ?? 0
    const newRetryCount = currentRetry + 1

    if (newRetryCount >= maxRetries) {
      const entry: PipelineStateEntry = {
        documentHash,
        stage: 'dead_letter',
        retryCount: newRetryCount,
        lastError: error,
        createdAt: existing?.createdAt ?? now,
        updatedAt: now,
      }
      await config.stateStore.save(entry)
      if (config.onTransition !== undefined && currentStage !== 'dead_letter') {
        config.onTransition(documentHash, currentStage, 'dead_letter', 'retry_exhausted')
      }
      return entry
    }

    const entry: PipelineStateEntry = {
      documentHash,
      stage: currentStage,
      retryCount: newRetryCount,
      lastError: error,
      createdAt: existing?.createdAt ?? now,
      updatedAt: now,
    }
    await config.stateStore.save(entry)
    return entry
  }

  /**
   * Reports whether a document has not yet exhausted its retry budget.
   */
  const shouldRetry = (entry: PipelineStateEntry): boolean => {
    return entry.retryCount < maxRetries
  }

  /**
   * Unconditionally marks a document as dead_letter regardless of retry count.
   */
  const markDeadLetter = async (
    documentHash: string,
    error: string,
  ): Promise<PipelineStateEntry> => {
    const existing = await config.stateStore.load(documentHash)
    const now = new Date().toISOString()
    const currentStage: PipelineStage = existing?.stage ?? 'received'

    const entry: PipelineStateEntry = {
      documentHash,
      stage: 'dead_letter',
      retryCount: existing?.retryCount ?? 0,
      lastError: error,
      createdAt: existing?.createdAt ?? now,
      updatedAt: now,
    }
    await config.stateStore.save(entry)

    if (config.onTransition !== undefined && currentStage !== 'dead_letter') {
      config.onTransition(documentHash, currentStage, 'dead_letter', 'mark_dead_letter')
    }
    return entry
  }

  /**
   * Returns all entries not in a terminal stage (indexed or dead_letter).
   */
  const listIncomplete = async (): Promise<readonly PipelineStateEntry[]> => {
    return config.stateStore.listIncomplete()
  }

  return {
    advance,
    recordFailure,
    shouldRetry,
    markDeadLetter,
    listIncomplete,
    maxRetries,
  } as const
}

File: go/ingest/state_machine.go
// SPDX-License-Identifier: Apache-2.0
package ingest

import (
	"context"
	"errors"
	"fmt"
	"sync"
	"time"
)

// PipelineStage represents the processing stage a document has reached
// within the ingestion pipeline.
type PipelineStage string

const (
	StageReceived  PipelineStage = "received"
	StageStored    PipelineStage = "stored"
	StageChunked   PipelineStage = "chunked"
	StageEmbedded  PipelineStage = "embedded"
	StageIndexed   PipelineStage = "indexed"
	StageCompleted PipelineStage = "completed"
	StageDeadLetter    PipelineStage = "dead_letter"
)

// stageOrder defines the ordinal position of each stage for comparison.
// Failed is excluded from the forward-progress chain.
var stageOrder = map[PipelineStage]int{
	StageReceived:  0,
	StageStored:    1,
	StageChunked:   2,
	StageEmbedded:  3,
	StageIndexed:   4,
	StageCompleted: 5,
}

// validForwardTransitions maps each stage to its single valid next stage.
var validForwardTransitions = map[PipelineStage]PipelineStage{
	StageReceived: StageStored,
	StageStored:   StageChunked,
	StageChunked:  StageEmbedded,
	StageEmbedded: StageIndexed,
	StageIndexed:  StageCompleted,
}

// TransitionEvent describes what triggered a state machine transition.
type TransitionEvent string

const (
	EventStageCompleted TransitionEvent = "stage_completed"
	EventStageFailed    TransitionEvent = "stage_failed"
	EventRetryExhausted TransitionEvent = "retry_exhausted"
)

// maxDefaultRetries is the default number of retries before a document
// moves to the dead_letter stage.
const maxDefaultRetries = 3

// ErrInvalidTransition signals that a requested stage transition is
// not permitted by the state machine rules.
var ErrInvalidTransition = errors.New("ingest: invalid state transition")

// PipelineStateEntry tracks the processing state of a single document
// through the ingestion pipeline.
type PipelineStateEntry struct {
	DocumentHash string
	Stage        PipelineStage
	RetryCount   int
	LastError    string
	CreatedAt    time.Time
	UpdatedAt    time.Time
}

// PipelineStateStore is the persistence layer for pipeline state entries.
// Implementations may back this with a database, file system, or in-memory
// map. P1-1 provides the production implementation.
type PipelineStateStore interface {
	// Load retrieves the state entry for a document hash. Returns nil if
	// no entry exists.
	Load(ctx context.Context, documentHash string) (*PipelineStateEntry, error)

	// Save persists the state entry, creating or overwriting as needed.
	Save(ctx context.Context, entry *PipelineStateEntry) error

	// ListIncomplete returns all entries not in a terminal stage
	// (completed or failed).
	ListIncomplete(ctx context.Context) ([]*PipelineStateEntry, error)
}

// TransitionCallback is invoked after every successful state transition.
// Implementations use this for observability (logging, metrics, events).
type TransitionCallback func(documentHash string, from, to PipelineStage, event TransitionEvent)

// PipelineStateMachine manages stage transitions for documents flowing
// through the ingestion pipeline. It validates transitions, tracks retry
// counts, and persists state via a PipelineStateStore.
type PipelineStateMachine struct {
	store      PipelineStateStore
	maxRetries int
	callback   TransitionCallback
	mu         sync.Mutex
}

// StateMachineConfig holds construction parameters for a PipelineStateMachine.
type StateMachineConfig struct {
	Store      PipelineStateStore
	MaxRetries int
	OnTransition TransitionCallback
}

// NewPipelineStateMachine creates a state machine with the given config.
// MaxRetries defaults to 3 if zero or negative.
func NewPipelineStateMachine(cfg StateMachineConfig) *PipelineStateMachine {
	retries := cfg.MaxRetries
	if retries <= 0 {
		retries = maxDefaultRetries
	}
	return &PipelineStateMachine{
		store:      cfg.Store,
		maxRetries: retries,
		callback:   cfg.OnTransition,
	}
}

// Transition validates that moving from current to the next stage implied
// by the event is permitted and returns the resulting stage. Returns
// ErrInvalidTransition for disallowed moves.
func Transition(current PipelineStage, event TransitionEvent) (PipelineStage, error) {
	switch event {
	case EventStageCompleted:
		next, ok := validForwardTransitions[current]
		if !ok {
			return current, fmt.Errorf("%w: no forward transition from %s", ErrInvalidTransition, current)
		}
		return next, nil
	case EventStageFailed:
		return current, nil
	case EventRetryExhausted:
		return StageDeadLetter, nil
	}
	return current, fmt.Errorf("%w: unknown event %s", ErrInvalidTransition, event)
}

// IsValidTransition reports whether moving from current to target is a
// valid single-step forward transition.
func IsValidTransition(current, target PipelineStage) bool {
	next, ok := validForwardTransitions[current]
	if !ok {
		return false
	}
	return next == target
}

// Advance loads the current state for documentHash, applies the event,
// and persists the resulting state. Returns the updated entry.
func (sm *PipelineStateMachine) Advance(ctx context.Context, documentHash string, event TransitionEvent) error {
	sm.mu.Lock()
	defer sm.mu.Unlock()

	entry, err := sm.store.Load(ctx, documentHash)
	if err != nil {
		return fmt.Errorf("ingest: loading state for %s: %w", documentHash, err)
	}
	if entry == nil {
		return fmt.Errorf("ingest: no state entry for document %s", documentHash)
	}

	from := entry.Stage
	next, transErr := Transition(entry.Stage, event)
	if transErr != nil {
		return transErr
	}

	entry.Stage = next
	entry.UpdatedAt = time.Now().UTC()

	if err := sm.store.Save(ctx, entry); err != nil {
		return fmt.Errorf("ingest: saving state for %s: %w", documentHash, err)
	}

	if sm.callback != nil && from != next {
		sm.callback(documentHash, from, next, event)
	}
	return nil
}

// ShouldRetry returns true if the entry has not yet exhausted its retry
// budget (retryCount < maxRetries).
func (sm *PipelineStateMachine) ShouldRetry(entry *PipelineStateEntry) bool {
	return entry.RetryCount < sm.maxRetries
}

// RecordFailure increments the retry counter. If retries are exhausted,
// the document transitions to the failed stage; otherwise it remains in
// the current stage for retry.
func (sm *PipelineStateMachine) RecordFailure(ctx context.Context, documentHash string, failErr error) error {
	sm.mu.Lock()
	defer sm.mu.Unlock()

	entry, err := sm.store.Load(ctx, documentHash)
	if err != nil {
		return fmt.Errorf("ingest: loading state for %s: %w", documentHash, err)
	}
	if entry == nil {
		return fmt.Errorf("ingest: no state entry for document %s", documentHash)
	}

	from := entry.Stage
	entry.RetryCount++
	entry.LastError = failErr.Error()
	entry.UpdatedAt = time.Now().UTC()

	if entry.RetryCount >= sm.maxRetries {
		entry.Stage = StageDeadLetter
		if err := sm.store.Save(ctx, entry); err != nil {
			return fmt.Errorf("ingest: saving state for %s: %w", documentHash, err)
		}
		if sm.callback != nil {
			sm.callback(documentHash, from, StageDeadLetter, EventRetryExhausted)
		}
		return nil
	}

	if err := sm.store.Save(ctx, entry); err != nil {
		return fmt.Errorf("ingest: saving state for %s: %w", documentHash, err)
	}
	if sm.callback != nil {
		sm.callback(documentHash, from, entry.Stage, EventStageFailed)
	}
	return nil
}

// MarkDeadLetter unconditionally moves the document to the dead_letter
// stage, recording the error regardless of retry count.
func (sm *PipelineStateMachine) MarkDeadLetter(ctx context.Context, documentHash string, failErr error) error {
	sm.mu.Lock()
	defer sm.mu.Unlock()

	entry, err := sm.store.Load(ctx, documentHash)
	if err != nil {
		return fmt.Errorf("ingest: loading state for %s: %w", documentHash, err)
	}
	if entry == nil {
		return fmt.Errorf("ingest: no state entry for document %s", documentHash)
	}

	from := entry.Stage
	entry.Stage = StageDeadLetter
	entry.LastError = failErr.Error()
	entry.UpdatedAt = time.Now().UTC()

	if err := sm.store.Save(ctx, entry); err != nil {
		return fmt.Errorf("ingest: saving state for %s: %w", documentHash, err)
	}
	if sm.callback != nil && from != StageDeadLetter {
		sm.callback(documentHash, from, StageDeadLetter, EventRetryExhausted)
	}
	return nil
}

// ListIncomplete returns all pipeline state entries that are not in a
// terminal stage (indexed/completed or dead_letter). Delegates to the
// underlying PipelineStateStore.
func (sm *PipelineStateMachine) ListIncomplete(ctx context.Context) ([]*PipelineStateEntry, error) {
	sm.mu.Lock()
	defer sm.mu.Unlock()
	return sm.store.ListIncomplete(ctx)
}

// V1PipelineStateEntry represents the pipeline state format from P0-1.
// It lacks retry tracking and uses a different field layout.
type V1PipelineStateEntry struct {
	DocumentID string
	Hash       string
	Stage      string // "stored", "chunked", "embedded", "indexed"
	UpdatedAt  time.Time
	ChunkCount int
}

// v1StageMap maps V1 stage strings to the V2 PipelineStage type.
var v1StageMap = map[string]PipelineStage{
	"stored":   StageStored,
	"chunked":  StageChunked,
	"embedded": StageEmbedded,
	"indexed":  StageIndexed,
}

// MigrateFromV1 converts a V1 pipeline state entry to the V2 format
// used by the state machine. V1 entries from P0-1 lack retry tracking.
func MigrateFromV1(v1 V1PipelineStateEntry) PipelineStateEntry {
	stage, ok := v1StageMap[v1.Stage]
	if !ok {
		stage = StageReceived
	}
	return PipelineStateEntry{
		DocumentHash: v1.Hash,
		Stage:        stage,
		RetryCount:   0,
		LastError:    "",
		CreatedAt:    v1.UpdatedAt,
		UpdatedAt:    v1.UpdatedAt,
	}
}
